# Choe et al (2020) Script Vignette


### For analyzing differential gene expression in R

Samples of the scripts used to analyze differential gene expression in [Choe et al. 2021] provided as an .Rmd. 

Full read count matrix and sample-to-library ID key provided as supplemental in the *DiB in preperation*. 
RNA-seq reads were aligned to assembly [TaeGut 3.2.4]: release 98 (now obsolete and replaced by [bTaeGut1_v1.p]: release 102).


You will need [R] and [RStudio], or another R compatible integrated development environment (IDE). 

Addtional files have been provided here to run the .Rmd vignette: 
  - Organized count matrix for Male vehicle Area X
  - List of activity dependent genes (ADGs) for all regions
  - Prior results for Male E2, Male exemestane, Female vehicle, Female E2, and Female exemestane Area X (run with TaeGut 3.2.4)
  - ensembl gene lists and biomart objects from TaeGut 3.2.4 

If there are any questions, contact [Ha Na Choe, PhD] or [Erich Jarvis, PhD]. 



   [Choe et al. 2021]: <https://doi.org/10.1016/j.yhbeh.2020.104911>
   [TaeGut 3.2.4]: <https://sep2019.archive.ensembl.org/Taeniopygia_guttata/Info/Index>
   [bTaeGut1_v1.p]: <https://uswest.ensembl.org/Taeniopygia_guttata/Info/Index?db=core>
   [R]: <https://cran.r-project.org/>
   [RStudio]: <https://rstudio.com/>
   [Ha Na Choe, PhD]: <mailto:ha.choe@duke.edu>
   [Erich Jarvis, PhD]: <mailto:ejarvis@rockefeller.edu>